### PR TITLE
prow plugin to block merge commits in a PR

### LINKED ISSUE #5376: Proposal: prevent merge commits via prow plugin
There's been a number of discussions around merge commits, and why they are bad. One of the more recent ones on kubernetes-dev@ is [here](https://groups.google.com/d/msg/kubernetes-dev/KmeNarMIyRU/5Z00p2WWCQAJ).

Right now it's up to the reviewers and approvers to notice that there are merge commits, and withhold lgtm/approve until the author squashes or rebases them away.

https://github.com/kubernetes/kubernetes/pull/51176 was opened to add a verify script to k/k to prevent them there, but I think this could be more easily done in a prow plugin (which would also let other repos easily turn it on).

For example, the GitHub API allows you to [list all the commits for a PR](https://developer.github.com/v3/pulls/#list-commits-on-a-pull-request), and count the number of "parents" of the commit. If it's >1, then it's a merge commit.

Example:
```
curl -s -H "Authorization: token ${GITHUB_API_TOKEN}" https://api.github.com/repos/kubernetes/community/pulls/1355/commits | jq '.[] | {sha: .sha, commit_message: .commit.message, number_of_parents: .parents | length}'
{
  "sha": "e4e04b4f25eb767c1b4e8c316c2a2c8ab2e30b3b",
  "commit_message": "sepell error",
  "number_of_parents": 1
}
{
  "sha": "574094e67d7c23beb23edfca014f0fca3e482979",
  "commit_message": "Merge remote-tracking branch 'community/master'",
  "number_of_parents": 2
}
{
  "sha": "964f90233cbb4550bc9ed00f62157b342cd7e58f",
  "commit_message": "spell error",
  "number_of_parents": 1
}
{
  "sha": "3d45caa2c516ecbb6fc7e689620bc35c4017ca72",
  "commit_message": "spell error",
  "number_of_parents": 1
}
```

If we set this up with a `do-not-merge/*` class label, then we could prevent auto-merging a PR containing a merge commit. The only issue would be if there is a legitimate case for a merge commit in a PR -- if this was the case, a branch manager or build cop would need to manually merge it as the SQ would refuse.

cc: @kubernetes/sig-contributor-experience-proposals @kubernetes/sig-testing-proposals @lavalamp @sttts @ixdy 
/area prow

### PR DESCRIPTION
**Pull request by Huang-Wei**
Add a prow plugin to block merge commits in a PR:

- implement MergeCommitsExistBetween() in git client
- add test cases in both githubClient and gitClient

Fixes #5376.

/cc @stevekuznetsov 

PS: it's my first PR for test-infra. Let me know if anything wrong.

### CODE REVIEW

File: prow/git/git_test.go [hunk 1]
```diff
@@ -133,3 +133,89 @@ func TestCheckoutPR(t *testing.T) {
 		t.Errorf("Didn't find file in PR after checking out: %v", err)
 	}
 }
+
+func TestMergeCommitsExistBetween(t *testing.T) {
+	lg, c, err := localgit.New()
```
Comment (MHBauer) [inline prow/git/git_test.go]: localgit lib looks impressive.

File: prow/git/localgit/localgit.go [hunk 2]
```diff
@@ -152,3 +152,21 @@ func (lg *LocalGit) RevParse(org, repo, commitlike string) (string, error) {
 	rdir := filepath.Join(lg.Dir, org, repo)
 	return runCmdOutput(lg.Git, rdir, "rev-parse", commitlike)
 }
+
+// MergeCommitsExistBetween does 'git log <target>..<head> --merged'.
+func (lg *LocalGit) MergeCommitsExistBetween(org, repo, target, head string) (string, error) {
```
Comment (stevekuznetsov) [inline prow/git/localgit/localgit.go]: With this your test is validating the `LocalGit` implementation but not the actual git client -- can we use the git client to run the check?

File: prow/plugins/mergecommitblocker/mergecommitblocker.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+// init registers out plugin as a pull request handler
+func init() {
+	plugins.RegisterPullRequestHandler(PluginName, handlePullRequest, helpProvider)
+}
+
+// helpProvider provides information on the plugin
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	// Only the Description field is specified because this plugin is not triggered with commands and is not configurable.
+	return &pluginhelp.PluginHelp{
+		Description: fmt.Sprintf(`The merge commit blocker plugin adds the do-not-merge/contains-merge-commit label to pull requests that contain merge commits`),
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker.go]: You are using `Sprintf` but have no formatting directives. Can we import the label name here and use the constant?

File: prow/plugins/mergecommitblocker/mergecommitblocker.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	defer func() {
+		if err := r.Clean(); err != nil {
+			log.WithError(err).Error("Error cleaning up repo.")
+		}
+	}()
+	if err := r.CheckoutPullRequest(num); err != nil {
+		return err
+	}
+	target, head := "origin/master", "HEAD"
+	// If we have specific SHAs, use them.
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker.go]: I think we are guaranteed to have those fields in the payload, and `origin/master` is assuming both the remote name and the target branch from the PR neither of which could be true, so I think we are OK to just use the `pre.PullRequest.{Base,Head}.SHA` values as-is

File: prow/plugins/mergecommitblocker/mergecommitblocker.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	if pre.PullRequest.Base.SHA != "" {
+		target = pre.PullRequest.Base.SHA
+	}
+	if pre.PullRequest.Head.SHA != "" {
+		head = pre.PullRequest.Head.SHA
+		if err := r.Checkout(pre.PullRequest.Head.SHA); err != nil {
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker.go]: You don't need to check out the SHA in question to do a `git log` for it, as long as you have the `SHA` fetched

File: prow/plugins/mergecommitblocker/mergecommitblocker.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	if err != nil {
+		return err
+	}
+	hasLabel := github.HasLabel(labels.MergeCommits, issueLabels)
+	if hasLabel && !existMergeCommits {
+		log.Infof("Removing %q Label for %s/%s#%d", labels.MergeCommits, owner, repo, num)
+		return ghc.RemoveLabel(owner, repo, num, labels.MergeCommits)
+	} else if !hasLabel && existMergeCommits {
+		log.Infof("Adding %q Label for %s/%s#%d", labels.MergeCommits, owner, repo, num)
+		return ghc.AddLabel(owner, repo, num, labels.MergeCommits)
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker.go]: Do we want to add a comment here to explain why the label is being added?

/cc @kubernetes/sig-contributor-experience-pr-reviews

File: prow/git/localgit/localgit.go (hunk 2 above)
Comment (Huang-Wei) [reply prow/git/localgit/localgit.go]: Absolutely, and `(lg *LocalGit) MergeCommitsExistBetween()` is removed.

File: prow/plugins/mergecommitblocker/mergecommitblocker.go (hunk 3 above)
Comment (Huang-Wei) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: Done.

File: prow/plugins/mergecommitblocker/mergecommitblocker.go (hunk 5 above)
Comment (Huang-Wei) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: Done.

File: prow/plugins/mergecommitblocker/mergecommitblocker.go (hunk 6 above)
Comment (Huang-Wei) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: Added a comment.

**Comment by Huang-Wei** [comment]
@stevekuznetsov Thanks for the review! Comments have been addressed. PTAL.

File: prow/plugins/mergecommitblocker/mergecommitblocker.go (hunk 6 above)
Comment (cblecker) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: I think what @stevekuznetsov was asking was should we add a comment to the issue itself, rather than a code comment.

An example would be the `blockade` plugin, that sets a comment on the issue when it adds the label, and prunes the comments when they no longer apply (https://github.com/kubernetes/test-infra/blob/ea3dad5c1be6a4b7e4cff6283519286e6adeaabb/prow/plugins/blockade/blockade.go#L162-L178)

Comment (Huang-Wei) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: Oops, hard to follow the "slang" here...

Updated. PTAL. Feel free to propose changes on "fixme" message.

File: prow/plugins/mergecommitblocker/mergecommitblocker.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+)
+
+const (
+	// PluginName defines this plugin's registered name.
+	PluginName = "mergecommitblocker"
+)
+
+var (
+	blockedBody = fmt.Sprintf("Adding label: `%s` because PR contains merge commits.", labels.MergeCommits)
+	fixmeBody   = "It seems merge commits were generated. Please eliminate them by rebasing the PR to master branch instead of merging master branch onto the PR."
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker.go]: This is good, but I think we could be a little more clear like this:

```
Your pull request contains merge commits, which are not allowed in this repository. Use `git rebase` to reapply your commits on top of the target branch.
```

File: prow/plugins/mergecommitblocker/mergecommitblocker.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+		if err := ghc.RemoveLabel(org, repo, num, labels.MergeCommits); err != nil {
+			return err
+		}
+		msg := plugins.FormatResponse(pre.PullRequest.User.Login, blockedBody, fixmeBody)
+		return ghc.CreateComment(org, repo, pre.Number, msg)
+	} else if !hasLabel && existMergeCommits {
+		// Adding this label to block a PR which has "merge" commits.
+		// To eliminate the label, PR author needs to rebase their PR onto
+		// master to flatten the git commits history.
+		// Ref: https://github.com/kubernetes/test-infra/issues/5376
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker.go]: Sorry for not being clear before -- I think we don't need this code comment as the code is now very clear.

File: prow/plugins/mergecommitblocker/mergecommitblocker.go (hunk 7 above)
Comment (Huang-Wei) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: Thanks! This is much more clear.

File: prow/plugins/mergecommitblocker/mergecommitblocker.go (hunk 8 above)
Comment (Huang-Wei) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: Done.

**Comment by Huang-Wei** [comment]
Commits have been squashed.

File: prow/plugins/mergecommitblocker/mergecommitblocker.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		return err
+	}
+	hasLabel := github.HasLabel(labels.MergeCommits, issueLabels)
+	if hasLabel && !existMergeCommits {
+		log.Infof("Removing %q Label for %s/%s#%d", labels.MergeCommits, org, repo, num)
+		if err := ghc.RemoveLabel(org, repo, num, labels.MergeCommits); err != nil {
+			return err
+		}
+		msg := plugins.FormatResponse(pre.PullRequest.User.Login, blockedBody, fixmeBody)
+		return ghc.CreateComment(org, repo, pre.Number, msg)
```
Comment (cjwagner) [inline prow/plugins/mergecommitblocker/mergecommitblocker.go]: I'm pretty sure this is backwards. We should post a comment when we add the label, not when we remove it (and similarly prune comments when we remove the label instead of when we add it).
Please validate this behavior with unit tests.

Comment (Huang-Wei) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: My bad on copying/pasting...

**Comment by Huang-Wei** [comment]
@cjwagner UT updated to cover github comments related logic. PTAL.

**Comment by Huang-Wei** [comment]
Thanks @cjwagner for the review!

**Comment by Huang-Wei** [comment]
Friendly ping @cblecker for a final review before cancel /hold.
